text/0000-convert-id.md

```diff
@@ -0,0 +1,201 @@
+- Feature Name: convert_id
+- Start Date: 2018-01-19
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Adds an identity function `pub fn id<T>(x: T) -> T { x }` as `core::convert::id`.
+The function is also re-exported to `std::convert::id` as well as the prelude of
+both libcore and libstd.
+
+# Motivation
+[motivation]: #motivation
+
+## The identity function is useful
+
+While it might seem strange to have a function that just returns back the input,
+there are some cases where the function is useful.
+
+### Using `id` to do nothing among a collection of mappers
+
+When you have collections such as maps or arrays of mapping functions like
+below and you watch to dispatch to those you sometimes need the identity
+function as a way of not transforming the input. You can use the identity
+function to achieve this.
+
+```rust
+// Let's assume that this and other functions do something non-trivial.
+fn do_interesting_stuff(x: u32) -> u32 { .. }
+
+// A dispatch-map of mapping functions:
+let mut map = HashMap::new();
+map.insert("foo", do_interesting_stuff);
+map.insert("bar", other_stuff);
+map.insert("baz", id);
+```
+
+### Using `id` as a no-op function in a conditional
+
+This reasoning also applies to simpler yes/no dispatch as below:
+
+```rust
+let mapper = if condition { some_manipulation } else { id };
+
+// do more interesting stuff inbetween..
+
+do_stuff(42);
+```
+
+### Using `id` to concatenate an iterator of iterators
+
+Given the law `join = (>>= id)`, we use the identity function to perform
```

Comment:
This sentence needlessly complicates a simple operation; I don't see any use of including the monad reference

Reply:
The operation `.flat_map(id)` is equivalent to join and may have some use for Haskllers who are reading. I can make the used syntax more Rusty tho.

Reply:
I mean, don't mention monad _at all_. Rust doesn't typically use that term. There is no reason to include it here. You're writing for Rustaceans, not Haskellers, just say "flatten" and "flat_map"

Reply:
I'm trying to choose my battles here =) I tried to come up with a compromise solution... But I don't like the *"don't mention the M word"* attitude in general.. Rustaceans say *"and_then"*, *"flat_map"*, etc... but they may also say *bind*, *"monad"*, and "join" at times since Rustaceans are a diverse group and some of them come from Haskell, Scala, F# and other such places. Also, I don't think that shying away from the word Monad helps learnability.

Reply:
No, that's the point. You aren't optimizing for Haskellers Rustaceans by using "monad". Haskeller Rustaceans know the Rust terminology perfectly well. You only end up disoptimizing for non-FP Rustaceans.

Rustaceans are a diverse group, but there's an option here that is adequate for _all_ Rustaceans.

This is literally the kind of thing that made folks confused by the pi types RFC. Now, this is at a much smaller scale, but I'm trying to stomp out the attitude of using unfamiliar terminology where not necessary in RFCs as much as possible.

Reply:
We have a fundamental disagreement here.

> Haskeller Rustaceans know the Rust terminology perfectly well.

Haskeller Rustaceans might not know that particular part of Rust... And nonetheless, there might be non-Rustacean functional programmers that might be interested in reading our RFCs or other documentation at some point. Besides, the current text of the RFC explains it in a way that is adequate for all Rustaceans since it includes the following language as well:
> In other words we are concatenating an iterator of iterators into a single iterator,

However, in an effort to not make this RFC block on this disagreement I'll agree to remove the language from the RFC.